123-job-jwt/proposal.md

```diff
@@ -0,0 +1,175 @@
+`auth_params` field in the `var_source.config.vault_config` schema should be
+re-used, then `auth_params: {role: "example", jwt: "((.CONCOURSE_CI_JWT))"}`
+might be a reasonable value, assuming that local environment variables can be
+referenced in such a way and we have pre-populated `CONCOURSE_CI_JWT` with the
+signed JWT during container pre-flight. However, this format would require more
+awareness on the part of the person configuring the pipeline, and the `jwt`
+field in the `auth_params` argument would always be the same, so it seems to
+make sense to abstract it in this case.
+
+Then, the call to the Vault API for the authentication step might look like the
+following:
+
+```sh
+#!/usr/bin/env bash
+auth_path=$(echo ${auth_path} | sed -E "s|^/*||g" | sed -E "s|/*$||g" |
+  sed -E "s|^auth/?||g" | sed -E "s|/?login$||g" | sed -E "s|/+|/|g"
+)
+resp=$(curl -s -X POST -H "X-Vault-Namespace: ${namespace}" \
+  --data @payload.json ${url}/v1/auth/${auth_path}/login
+)
+export VAULT_TOKEN=$(echo "${resp}" | jq -r ".auth.client_token")
+```
+
+or (forgive my golang pseudocode, rusty and untested):
+
+```golang
+// not dealing with errors or worrying about syntax, this is pseudocode
+// but please still make fun of my pseudocode, maybe I'll learn something
+func (ap *VarSources.Config.auth_path) FmtAuthPath() nil {
+  ap := strings.Trim(ap, "/")
+  ap := strings.TrimPrefix(ap, "auth/")
+  ap := strings.TrimSuffix(ap, "/login")
+  ap := path.Clean(ap)
+}
+type LoginData struct {
+  Role string `json:"role"`
+  JWT  string `json:"jwt"`
+}
+type LoginResp struct {
+  auth struct {
+    ClientToken string   `json:"client_token"`
+    Accessor    string   `json:"accessor"`
+    Policies    []string `json:"policies"`
+    Lease       int      `json:"lease_duration"`
+    Renewable   bool     `json:"renewable"`
+  } `json:"auth"`
+}
+func (c *VarSources.Config) VaultAuth() (token string, accessor string, err error) {
+  // I'm sure the vault package implements the auth method much better than this
+  client := &http.Client{}
+  uri := fmt.Sprintf("%s/v1/auth/%s/login", c.url, c.auth_path.FmtAuthPath())
+  jsonData := LoginData{
+    Role: c.auth_role,
+    JWT:  os.Getenv("CONCOURSE_CI_JWT")
+  }
+  var buf bytes.Buffer
+  err := json.NewEncoder(&buf).Encode(jsonData)
+  req, err := http.NewRequest("POST", uri, &buf)
+  req.Header.Add("X-Vault-Namespace", c.namespace)
+  resp, err := client.Do(req)
+  defer resp.Body.Close()
+  var dec LoginResp
+  dec = json.NewDecoder(&buf).Decode(resp)
+  return dec.auth.ClientToken, dec.auth.Accessor, nil
+}
+```
+
+## Open Questions
+
+- Where should the claims be aggregated and signing portion be completed? (atc?)
+- How should the signed JWT be delivered to the container runtime?
+- Where should the validation endpoint live? (web?)
+- Where should the signed JWT live? (Environment Variable + File?)
+- Should the validation step fail after the container dies regardless of `exp`
```

Comment:
This is more talking about the behaviour of the validation endpoint.

Generally, the `jwks_url="https://concourse-ci.example.internal/web/-/jwks"` endpoint would only validate the signature and say it's valid. However, in the Concourse use case, if the Job, Step, or Pipeline has finished, should this endpoint still return a "valid JWT" kind of response? In my opinion, it should not, because the context for the JWT no longer exists, which generally means that the JWT is being leveraged by something other than the pipeline/job/step, which would be unintended. So we would want to write additional logic on the `jwks` endpoint that also checks the pipeline/job/step IDs inside the header or payload of the JWT and ensure that they are all still in an active state by checking with `atc`.

However, we may want to make this a behaviour that can be overridden with an env var on the `web` node(s) (perhaps `ALLOW_JWT_FROM_EXPIRED_JOBS=true` or similar) for the purposes of development or debugging?

Reply:
This probably depends on the context of the validator - I assume this "validation endpoint" is only used inside concourse. I'd expect outsiders to only care about `/.well-known/jwks.json` and the `exp` claim.

Reply:
Yes, you're correct. This endpoint doesn't actually do any processing, and is simply a discovery URL so that the application can retrieve the latest `jwks.json` data which contains the public asymmetric key (among other things). In my earlier response, I was looking at older documentation that used a symmetric key, and so my earlier response in this thread is incorrect.